### PR TITLE
Fix None slice bounds with strict-optional

### LINKED ISSUE #3442: Slice index false positive
```
from typing import Union, Any

d: Union[int, None]
q: Any
q[0:d]  # error: Slice index must be an integer or None
```

Reported by @chrish42.

### PR DESCRIPTION
**Pull request by miedzinski**
Fixes #3442.

### CODE REVIEW

File: test-data/unit/check-expressions.test [hunk 1]
```diff
@@ -998,6 +998,7 @@ a[:o]  # E: Slice index must be an integer or None
 [builtins fixtures/slice.pyi]
 
 [case testNoneSliceBounds]
+# flags: --strict-optional
```
Comment (JelleZijlstra) [inline test-data/unit/check-expressions.test]: We should probably test this both with and without strict-optional.

File: mypy/checkexpr.py [hunk 2]
```diff
@@ -1893,10 +1893,11 @@ def analyze_super(self, e: SuperExpr, is_lvalue: bool) -> Type:
             return AnyType()
 
     def visit_slice_expr(self, e: SliceExpr) -> Type:
+        expected = UnionType.make_union([NoneTyp(), self.named_type('builtins.int')])
```
Comment (gvanrossum) [inline mypy/checkexpr.py]: Should probably use `typeanal.make_optional_type()` instead.

Comment (miedzinski) [inline mypy/checkexpr.py]: I was looking for such function/method, but couldn't find it. Thanks!

**Comment by gvanrossum** [comment]
Thanks!

**Pull request merged**